### PR TITLE
Vector icons

### PR DESCRIPTION
**Pull request by ztNFny**
Hi,
as discussed during the dev call here's the PR for vector icons before I go on vacation. Don't merge!

Known open topics:
- [ ] all places that use icons need to build the icon from multiple layers (filter dialog, search dropdown, ...). There'll probably some slight difference in the icon display depending on the place: Map = with pin, list/search = with (circle) marker but without pin, filter dialog = without marker (should be enough, needs to be tested)
- [x] icons need to be centered properly (failing today due to rounding errors), see https://github.com/cgeo/cgeo/issues/11139#issuecomment-894852597 and  #11231
-> This is partly solved by https://github.com/cgeo/cgeo/pull/11454/commits/637ea3681642644d6cf11ce2ece7cf87df7dfa01 but a proper solution for SDK >= 23 using setLayerGravity should be implemented (see https://github.com/cgeo/cgeo/issues/11139#issuecomment-896660188 ff)
- [x] "Select item" menu (for selecting a cache from map) doesn't have enough room for the new map markers
- [ ] #11424
- [x] for visited waypoints add an overlay/marker instead of the transparent marker
- [x] remove dead code for dot_black -> not sure about that one, it's currently used as a reference for the size of custom markers (btw: I knowingly display the markers in upper left and lower left corner too large for dot_black to fit inbetween - if I make them smaller they're just too tiny. As all 3 left-column icons are rarely displayed together this shouldn't be too much of an issue - if you think it is then let's discuss when I'm back)
- [ ] dot markers should also change shape -> I'll take care of that
- [ ] single waypoint map (called by selecting navigation target in waypoint list) needs to know the gccode to be able to display the correct marker shape (fixed already for OSM, but GM needs work)

During testing I also sometimes - completly unrepoducable however but by sheer luck - get some display issues. Like the upper half of a specific icon type missing, or the marker pin being colored (it should always be white).
Maybe MapMarkerUtils is getting too crowded and needs to be rewritten in parts for performance - might make sense to implement "all places that use icons need to build the icon from multiple layers" anyway.

Please don't touch the drawables themselves, I'm still working on most of them.

### CODE REVIEW

**Comment by Lineflyer** [comment]
add to whitelist

**Comment by moving-bits** [comment]
@ztNFny 
Thanks for providing the draft. Can you rebase it to the current state of `master` branch of c:geo so that the conflicts mentioned above are resolved? Would help us in providing updates to your PR and with interpreting Codacy / build status info.

**Comment by UniQP** [comment]
@ztNFny I rebased the branch to `master` (saw moving-bits' comment too late, I hope you hadn't started a rebase yourself at the same time).

**Comment by moving-bits** [comment]
Thanks @UniQP for rebasing. Can you share the git command how you did this for another user's repo? I'd like to put that command into my "git cheat sheet" :-)

**Comment by UniQP** [comment]
@moving-bits
1. Add the fork of the user as a remote: `git remote add ztNFny git@github.com:ztNFny/cgeo`
2. Fetch changes (of all remotes): `git fetch --all`
3. Switch to the branch: `git switch vectorIcons`
4. Make local changes and push them as usual.

**Comment by moving-bits** [comment]
Thanks @UniQP, will add that to my git cheat sheet.
BTW: I'm using `https://github.com/ztNFny/cgeo` instead of `git@github.com:ztNFny/cgeo`

**Comment by moving-bits** [comment]
> When viewing all waypoints of a cache they have the correct icon. But if you select only a single waypoint (go to waypoints list, then click the compass icon next to the waypoint) it get's the hexagon background - it seems the connector is not properly set at that stage

Geocode is available in the waypoint list, so the connector could be determined, but clicking on the navigation item does a `DefaultMap.startActivityCoords`, which does not pass the geocode with it. If I change `startActivityCoords` to pass the geocode, map shows the whole geocache with all its waypoints... :-(

Need to investigate further.

**Comment by moving-bits** [comment]
I've found a solution for OSM maps. Cannot test with GM, though, as GM currently is not working for me locally.

**Comment by ztNFny** [comment]
> I've found a solution for OSM maps. Cannot test with GM, though, as GM currently is not working for me locally.

Not working for GM, ok for all the OSM maps.

**Comment by moving-bits** [comment]
> * [x]  "Select item" menu (for selecting a cache from map) doesn't have enough room for the new map markers

I've added a "minHeight" rule for now, but a refactoring should be done probably to use a separate `ImageView` for the cache marker and adapt the code in the calling places.

**Comment by moving-bits** [comment]
> > I've found a solution for OSM maps. Cannot test with GM, though, as GM currently is not working for me locally.
> 
> Not working for GM, ok for all the OSM maps.

I've extended the TODO list accordingly so that this item does not get lost

**Comment by moving-bits** [comment]
> > I've found a solution for OSM maps. Cannot test with GM, though, as GM currently is not working for me locally.
> 
> Not working for GM, ok for all the OSM maps.

Someone else will have to look into it for Google Maps.

I've spend a couple of hours trying to get Google Maps back running in the emulator, but to no avail.
All I get are some cryptic error messages, which lead to nothing when searching for them... 😠 

(The problem arose "out of the blue" some months ago, and even back at that time I did not now what I may have changed in configuration or else.)

**Comment by moving-bits** [comment]
added icon positioning for API 23+ using `setLayerGravity`, while staying with our current method for API < 23

**Comment by moving-bits** [comment]
The commit history looks a bit weird to me, e. g. see the whole bunch of "Crowdin" commits here: https://ci.cgeo.org/job/cgeo%20pull%20request/6815/

- Can you rebase to current upstream master again - should get rid of the failing test and hopefully fix the commit history,
- probably squashing the last merges along the way (rebasing should not result in new commits), 
- and then set the PR's status to "ready for review" (removing the "draft" status)?

**Comment by ztNFny** [comment]
done

**Comment by fm-sys** [comment]
Still not working. Problem is that you probably have merge cgeo master into your branch instead of doing a rebase. So probably it would be the best to just drop the merge commits from the branch. 

Then rebase as follow:
`git pull --rebase upstream master` 

Merging is currently not possible:
Screenshot_20210828_221233_org mozilla firefox

But squashing via git UI whould somehow work. I would be fine with squashing and merging the current state via the web UI. At least this would be the easiest way... (and no fiddling around with git ;-)) 
Screenshot_20210828_221853_org mozilla firefox

**Comment by ztNFny** [comment]
there were some conflicts that prevented a rebase, I had to merge
You should have write acess to the repo via this PR if you want to try to fix something yourself.#!/bin/sh
workdir=/home/lars/.adticket

fetchandfind() {
	groupid=$1
	filename=$2
	
	echo fetching $filename
	
	wget -q --no-check-certificate "https://www.adticket.de/Erweitertesuche.html?gid=${groupid}&limit=0&search_city=Mainz&search_end_date=&search_location=&search_name=&search_radius=100&search_start_date=&search_string=&start=0" -O $workdir/e.txt
	mv $workdir/e-${filename}.txt $workdir/e-${filename}-old.txt
	# 1. Whitespace raus
	# 2. Whitespace raus
	# 3. Whitespace gezielt wieder rein
	# 4. Whitespace gezielt wieder rein
	# 5. Gezielt die Zeilen mit den eigentlichen Ergebnissen suchen
	cat $workdir/e.txt | tr -d '\t' | tr -d '\n' | sed 's/<\/tbody>/\n<\/tbody>/g' | sed -r 's/<tr onclick="javascript:window.document.location ='\''([^>]*)'\''"/\n<tr><td><a href="https:\/\/www.adticket.de\1">Link<\/a/g' | grep '<tr><td><a href' | sort | uniq > $workdir/e-${filename}.txt
	
	echo \<tr\>\<th colspan="6"\>${filename}\</th\> >> $workdir/e-diff.html
	#grep -i -f artistsdiv.txt $workdir/e-${filename}.txt >> debug1.html
	#grep -i -f artistsdiv.txt $workdir/e-${filename}.txt | grep -i --color=always -f artists.txt >> debug2.html
	grep -i -f $workdir/artistsdiv.txt $workdir/e-${filename}.txt | grep -i --color=always -f $workdir/artists.txt | sed $'s/\\(<div>[^<]*\\)\x1b\\[01;31m\x1b\\[K\\([^<]*\\)\x1b\\[m\x1b\\[K\\([^<]*<\\/div>\\)/\\1<span style="color:red">\\2<\\/span>\\3/g' | sed $'s/\x1b\\[[0-9;]*[a-zA-Z]//g' >> $workdir/e-diff.html
}

#rm debug1.html
#rm debug2.html
echo \<head\>\<base href=\"http://www.eventim.de/\"\>\</head\>\<body\>\<table border=1\> > $workdir/e-diff.html

fetchandfind "3,15" "musical"
fetchandfind "2,2" "rockpop"
fetchandfind "2,4" "metal"

echo \</table\>\</body\> >> $workdir/e-diff.html

cat $workdir/mailheader.txt $workdir/e-diff.html | sendmail -t

currentdate=`date +%s`
mkdir $workdir/$currentdate
cp $workdir/e*.txt $workdir/$currentdate

**Comment by fm-sys** [comment]
> there were some conflicts that prevented a rebase, I had to merge
> You should have write acess to the repo via this PR if you want to try to fix something yourself.

Somehow it must be possible, but probably it's not worth the hassle. If no one stops me, I would suggest simply doing a "squash-merge" tomorrow... Did I understood it right, that we want merge the PR to master now?

**Comment by ztNFny** [comment]
> Did I understood it right, that we want merge the PR to master now?

That's what we discussed in the call 2 weeks ago.
There's still a lot of open items, but I hope they get fixed faster in master. I'll focus on finalizing the icons instead of code changes.

**Comment by fm-sys** [comment]
> > Did I understood it right, that we want merge the PR to master now?
> 
> That's what we discussed in the call 2 weeks ago.

Unfortunately I had no time to join on that day, but this sounds like a very good plan. Thanks a lot for this create contribution to c:geo!

**Comment by Lineflyer** [comment]
> Did I understood it right, that we want merge the PR to master now?

Actually, lets not do it right now but wait some days after todays release has proven stable, then I would like to forward `release` to head of `master`, in preparation for next beta.
Then `master` would be all open for WIP items.

**Comment by fm-sys** [comment]
> I would like to forward `release` to head of `master`, in preparation for next beta.

Yes, good idea to make a cut before starting with new larger topics on `master`. I will wait with merging...

**Comment by bekuno** [comment]
There are some checkstyle problems:
~~~
15:54:17  [ant:checkstyle] [WARN] /home/jenkins/slave/workspace/cgeo pull request/main/src/cgeo/geocaching/utils/EmojiUtils.java:31:8: Unused import - android.widget.Button. [UnusedImports]
15:54:17  [ant:checkstyle] [INFO] /home/jenkins/slave/workspace/cgeo pull request/main/src/cgeo/geocaching/utils/MapMarkerUtils.java:132:103: 'typecast' is not followed by whitespace. [WhitespaceAfter]
15:54:17  [ant:checkstyle] [INFO] /home/jenkins/slave/workspace/cgeo pull request/main/src/cgeo/geocaching/utils/MapMarkerUtils.java:132:119: '*' is not followed by whitespace. [WhitespaceAround]
15:54:17  [ant:checkstyle] [INFO] /home/jenkins/slave/workspace/cgeo pull request/main/src/cgeo/geocaching/utils/MapMarkerUtils.java:132:119: '*' is not preceded with whitespace. [WhitespaceAround]
15:54:17  [ant:checkstyle] [INFO] /home/jenkins/slave/workspace/cgeo pull request/main/src/cgeo/geocaching/utils/MapMarkerUtils.java:353:16: Variable 'geocode' should be declared final. [FinalLocalVariable]
~~~

You can check them before in AS using the checkstyle-IDEA plugin and in the commit dialog automatic when activating the button "Scan with checkstyle".

**Comment by moving-bits** [comment]
@ztNFny 
I'm preparing a new beta release currently. Are you fine with merging your PR to `master` after beta preparations are finished (meaning: probably tonight)?

**Comment by ztNFny** [comment]
> @ztNFny
> I'm preparing a new beta release currently. Are you fine with merging your PR to `master` after beta preparations are finished (meaning: probably tonight)?

I rebased it earlier today, should be good to go.

File: main/src/cgeo/geocaching/utils/MapMarkerUtils.java [hunk 1]
```diff
@@ -219,88 +347,76 @@ private static void addListMarkers(final Resources res, final InsetsBuilder inse
...
      */
-    @NonNull
-    @SuppressWarnings("PMD.NPathComplexity") // method readability will not improve by splitting it up
-    private static LayerDrawable createCacheMarker(final Resources res, final Geocache cache, @Nullable final CacheListType cacheListType, final ArrayList<Integer> assignedMarkers) {
-        final int useEmoji = cache.getAssignedEmoji();
-
-        // background: disabled or not
-        final Drawable marker = ResourcesCompat.getDrawable(res, cache.getMapMarkerId(), null);
-        final InsetsBuilder insetsBuilder = new InsetsBuilder(res, marker.getIntrinsicWidth(), marker.getIntrinsicHeight());
+    public static LayerDrawable createCacheTypeMarker(final Resources res, final CacheType type) {
```
Comment (fm-sys) [inline main/src/cgeo/geocaching/utils/MapMarkerUtils.java:402]: Probably, this method should be made private as it is designed for internal usage only while the other method directly above (with caching included) is designed for usage in cgeo.

Comment (ztNFny) [reply main/src/cgeo/geocaching/utils/MapMarkerUtils.java]: > Probably, this method should be made private as it is designed for internal usage only while the other method directly above (with caching included) is designed for usage in cgeo.

done. I also added caching for dotMarkers while at it

**Pull request merged**